File: tools/check_screen_registry.py
#!/usr/bin/env python3
"""
check_screen_registry.py — CI guard for the screen <-> _EditorState registry.

Purpose
-------
Every screen in a canvas app must be registered in the app's editor state
(_EditorState.pa.yaml, ScreensOrder list). A screen file that exists in
src/Src/ but is missing from _EditorState never gets loaded by the app; an
_EditorState entry with no matching file breaks the pack. This script fails
(exit code 1) on either mismatch so CI catches the drift immediately.

Checks (both directions)
------------------------
1. Screen file exists (src/Src/scr_*.pa.yaml) but has no matching
   ScreensOrder entry in _EditorState.pa.yaml  -> FAIL
2. ScreensOrder entry in _EditorState.pa.yaml with no matching screen file
   in src/Src/                                 -> FAIL

Usage
-----
    python3 tools/check_screen_registry.py [--src DIR] [--editor FILE]

Defaults to the repo layout (src/Src/ + src/Src/_EditorState.pa.yaml).
Exit code 0 = registry consistent, 1 = mismatches found.
"""

import argparse
import glob
import os
import re
import sys

# Repo-relative defaults (script lives in tools/, repo root is one level up)
DEFAULT_SRC = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "src", "Src")
DEFAULT_EDITOR = os.path.join(DEFAULT_SRC, "_EditorState.pa.yaml")


def screen_name_from_file(path):
    """
    Return the screen name declared inside the file's `Screens:` block.
    Prefer the declared key over the filename so renames stay in sync.
    """
    with open(path, encoding="utf-8-sig") as f:
        head = f.read(2000)
    # The file's first block is `Screens:` followed by `  <ScreenName>:`.
    m = re.search(r"^Screens:\s*\n\s*([A-Za-z_][A-Za-z0-9_]*):", head, re.M)
    if m:
        return m.group(1)
    # Fallback: infer from the filename (scr_Foo.pa.yaml -> scr_Foo).
    base = os.path.basename(path)
    if base.endswith(".pa.yaml"):
        return base[: -len(".pa.yaml")]
    return None


def editor_screens(path):
    """Return the ordered set of screen names listed under ScreensOrder."""
    screens = []
    if not os.path.exists(path):
        return screens
    with open(path, encoding="utf-8-sig") as f:
        content = f.read()
    in_order = False
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.startswith("ScreensOrder:"):
            in_order = True
            continue
        if in_order:
            # List items look like "- scr_Home"; stop at the next top-level key.
            m = re.match(r"^-\s*([A-Za-z_][A-Za-z0-9_]*)\s*$", stripped)
            if m:
                screens.append(m.group(1))
            elif stripped and not stripped.startswith("-"):
                in_order = False
    return screens


def main():
    parser = argparse.ArgumentParser(description="CI check: screen files <-> _EditorState registry")
    parser.add_argument("--src", default=DEFAULT_SRC, help="Directory containing *.pa.yaml screen files")
    parser.add_argument("--editor", default=DEFAULT_EDITOR, help="Path to _EditorState.pa.yaml")
    args = parser.parse_args()

    src_dir = os.path.abspath(args.src)
    editor_path = os.path.abspath(args.editor)

    if not os.path.isdir(src_dir):
        print(f"ERROR: screen source directory not found: {src_dir}")
        sys.exit(1)
    if not os.path.exists(editor_path):
        print(f"ERROR: _EditorState.pa.yaml not found: {editor_path}")
        sys.exit(1)

    # ---- collect screen files (exclude App.pa.yaml and _EditorState itself) ----
    screen_files = {}
    for path in sorted(glob.glob(os.path.join(src_dir, "*.pa.yaml"))):
        base = os.path.basename(path)
        if base in ("App.pa.yaml", "_EditorState.pa.yaml"):
            continue
        name = screen_name_from_file(path)
        screen_files[name] = base  # name -> filename

    registered = set(editor_screens(editor_path))

    # ---- direction 1: file exists but not registered ----
    unregistered = sorted(set(screen_files) - registered)
    # ---- direction 2: registered but no file ----
    orphaned = sorted(registered - set(screen_files))

    problems = []
    if unregistered:
        problems.append(
            "Screen file(s) exist but are MISSING from _EditorState ScreensOrder: "
            + ", ".join(f"{screen_files[n]} ({n})" for n in unregistered)
        )
    if orphaned:
        problems.append(
            "_EditorState ScreensOrder entry(ies) with NO matching screen file: "
            + ", ".join(orphaned)
        )

    # ---- report ----
    print("=" * 72)
    print("Screen registry check — src/Src/*.pa.yaml <-> _EditorState.pa.yaml")
    print("=" * 72)
    print(f"Screen files found : {len(screen_files)}")
    print(f"EditorState screens: {len(registered)}")
    for p in problems:
        print("FAIL:", p)
    if not problems:
        print("OK: every screen file is registered and every entry has a file.")
    print("-" * 72)
    print(f"Problems: {len(problems)}")
    sys.exit(1 if problems else 0)


if __name__ == "__main__":
    main()

File: tools/check_screen_registry.mjs
#!/usr/bin/env node
/**
 * check_screen_registry.mjs — Node port of tools/check_screen_registry.py.
 *
 * Keeps the unpacked canvas-app source honest without Python, so the
 * Freebuff preview build step (which runs on a Node-only image) can guard
 * the same invariant as the repo's CI job:
 *
 *   1. A screen file exists (src/Src/scr_*.pa.yaml) but has no matching
 *      ScreensOrder entry in _EditorState.pa.yaml  -> FAIL
 *   2. A ScreensOrder entry has no matching screen file                  -> FAIL
 *
 * Usage:
 *   node tools/check_screen_registry.mjs [--src DIR] [--editor FILE]
 *
 * Defaults match the repo layout (src/Src + src/Src/_EditorState.pa.yaml).
 * Exit code 0 = registry consistent, 1 = mismatches found.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const TOOLS_DIR = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(TOOLS_DIR, "..");
const DEFAULT_SRC = path.join(REPO_ROOT, "src", "Src");
const DEFAULT_EDITOR = path.join(DEFAULT_SRC, "_EditorState.pa.yaml");

function parseArgs(argv) {
  const args = { src: DEFAULT_SRC, editor: DEFAULT_EDITOR };
  for (let i = 2; i < argv.length; i += 1) {
    if (argv[i] === "--src" && argv[i + 1]) {
      args.src = argv[i + 1];
      i += 1;
    } else if (argv[i] === "--editor" && argv[i + 1]) {
      args.editor = argv[i + 1];
      i += 1;
    }
  }
  return args;
}

function readText(p) {
  // Mirror Python's utf-8-sig: strip a leading BOM if present.
  let s = fs.readFileSync(p, "utf8");
  if (s.charCodeAt(0) === 0xfeff) s = s.slice(1);
  return s;
}

/** Screen name declared inside the file's `Screens:` block (first key). */
function screenNameFromFile(p) {
  const head = readText(p).slice(0, 2000);
  const m = head.match(/^Screens:\s*\n\s*([A-Za-z_][A-Za-z0-9_]*):/m);
  if (m) return m[1];
  const base = path.basename(p);
  return base.endsWith(".pa.yaml") ? base.slice(0, -".pa.yaml".length) : null;
}

/** Ordered list of screen names under `ScreensOrder:` in _EditorState. */
function editorScreens(p) {
  if (!fs.existsSync(p)) return [];
  const screens = [];
  let inOrder = false;
  for (const raw of readText(p).split("\n")) {
    const line = raw.trim();
    if (line.startsWith("ScreensOrder:")) {
      inOrder = true;
      continue;
    }
    if (!inOrder) continue;
    const m = line.match(/^-\s*([A-Za-z_][A-Za-z0-9_]*)\s*$/);
    if (m) {
      screens.push(m[1]);
    } else if (line.length > 0 && !line.startsWith("-")) {
      inOrder = false; // next top-level key
    }
  }
  return screens;
}

const args = parseArgs(process.argv);
const srcDir = path.resolve(args.src);
const editorPath = path.resolve(args.editor);

if (!fs.existsSync(srcDir) || !fs.statSync(srcDir).isDirectory()) {
  console.error(`ERROR: screen source directory not found: ${srcDir}`);
  process.exit(1);
}
if (!fs.existsSync(editorPath)) {
  console.error(`ERROR: _EditorState.pa.yaml not found: ${editorPath}`);
  process.exit(1);
}

const screenFiles = {}; // name -> filename
for (const name of fs.readdirSync(srcDir).sort()) {
  if (!name.endsWith(".pa.yaml")) continue;
  if (name === "App.pa.yaml" || name === "_EditorState.pa.yaml") continue;
  const p = path.join(srcDir, name);
  screenFiles[screenNameFromFile(p)] = name;
}

const registered = new Set(editorScreens(editorPath));

const unregistered = Object.keys(screenFiles)
  .filter((n) => !registered.has(n))
  .sort();
const orphaned = [...registered]
  .filter((n) => !(n in screenFiles))
  .sort();

const problems = [];
if (unregistered.length > 0) {
  problems.push(
    "Screen file(s) exist but are MISSING from _EditorState ScreensOrder: " +
      unregistered.map((n) => `${screenFiles[n]} (${n})`).join(", ")
  );
}
if (orphaned.length > 0) {
  problems.push(
    "_EditorState ScreensOrder entry(ies) with NO matching screen file: " +
      orphaned.join(", ")
  );
}

console.log("=".repeat(72));
console.log("Screen registry check — src/Src/*.pa.yaml <-> _EditorState.pa.yaml");
console.log("=".repeat(72));
console.log(`Screen files found : ${Object.keys(screenFiles).length}`);
console.log(`EditorState screens: ${registered.size}`);
for (const p of problems) console.log("FAIL:", p);
if (problems.length === 0) {
  console.log("OK: every screen file is registered and every entry has a file.");
}
console.log("-".repeat(72));
console.log(`Problems: ${problems.length}`);
process.exit(problems.length > 0 ? 1 : 0);
